flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java

```diff
@@ -103,14 +105,41 @@ public void initializeState(StateInitializationContext context) throws Exception
     if (context.isRestored()) {
       if (globalStatisticsState.get() == null
           || !globalStatisticsState.get().iterator().hasNext()) {
-        LOG.warn(
+        LOG.info(
             "Operator {} subtask {} doesn't have global statistics state to restore",
             operatorName,
             subtaskIndex);
+        // If Flink deprecates union state in the future, RequestGlobalStatisticsEvent can be
+        // leveraged to request global statistics from coordinator if new subtasks (scale-up case)
+        // has nothing to restore from.
       } else {
-        LOG.info(
-            "Operator {} subtask {} restoring global statistics state", operatorName, subtaskIndex);
-        this.globalStatistics = globalStatisticsState.get().iterator().next();
+        AggregatedStatistics restoredStatistics = globalStatisticsState.get().iterator().next();
+        // Range bounds is calculated from reservoir samplings with the determined partition
+        // (downstream parallelism). If downstream parallelism changed due to rescale, the computed
+        // range bounds array is not applicable. Operators should request coordinator to recompute
+        // the range bounds array using the new parallelism.
+        if (restoredStatistics.type() == StatisticsType.Sketch
+            && restoredStatistics.keySamples().length + 1 != downstreamParallelism) {
+          LOG.info(
+              "Operator {} subtask {} ignored restored sketch statistics as range bound length "
+                  + "not matching downstream parallelism due to rescale. "
+                  + "Old parallelism is {}, new parallelism is {}",
+              operatorName,
+              subtaskIndex,
+              restoredStatistics.keySamples().length + 1,
+              downstreamParallelism);
+          // Asynchronously request the latest global statistics calculated with new downstream
+          // parallelism. It is possible events may have started flowing before coordinator responds
+          // with global statistics. In this case, range partitioner would just blindly shuffle
+          // records in round robin fashion.
```

Comment:
Maybe we could be better off if we use the old ranges with some heuristics?
- If we have higher number of subtasks, just use the old ranges, and leave idle tasks
- If we have lower number of subtasks, then use modulo?

Or the communication is fast enough, that it doesn't worth the complexity?

Reply:
great question. I also thought about the fallback heuristic. Initially, I was thinking maybe start with simple and we can improve this part if it turns out to be a problem. communication should be relatively fast (maybe a few to 10 ms) if parallelism/fan-out is not very high.

> If we have higher number of subtasks, just use the old ranges, and leave idle tasks

scale-up doesn't require any code change. it works like this already

> If we have lower number of subtasks, then use modulo?

I agree modulo could be a sensible strategy
* pro: better clustering than round-robin
* con: uneven distribution. some subtasks may get double the loads than the other subtasks. but if the stats refresh is fast (like less than a dozen of ms). maybe this is not a concern.

Hence, I am in favor of implementing the fallback behavior for rescale

Reply:
BTW, we haven't added the SketchRangePartitioner yet. fallback handling would be implemented there. so it will be out of the scope for this PR.

Reply:
added the SketchRangePartitioner and RangePartitioner to this PR. so the scope is bigger now